lib/solver.ml

```diff
@@ -65,25 +70,112 @@ let rec solve_eq sol ty1 ty2 =
   | (TyStruct tys1, TyStruct tys2) when List.length tys1 = List.length tys2 ->
      let open Result in
      List.zip_exn tys1 tys2
-     |> List.fold_left ~init:(Ok sol) ~f:(fun res (ty1,ty2) -> res >>= fun sol -> solve_eq sol ty1 ty2) 
+     |> List.fold_left ~init:(Ok sol) ~f:(fun res (ty1,ty2) -> res >>= fun sol -> solve_eq sol ty1 ty2)
   | (TyStruct tys1, TyStruct tys2) ->
      Error (Failure ("the tuple types are not different length"))
   | (TyFun (tys1, ty1), TyFun (tys2, ty2)) when List.length tys1 = List.length tys2 ->
      let open Result in
      List.zip_exn (ty1::tys1) (ty2::tys2)
-     |> List.fold_left ~init:(Ok sol) ~f:(fun res (ty1,ty2) -> res >>= fun sol -> solve_eq sol ty1 ty2) 
+     |> List.fold_left ~init:(Ok sol) ~f:(fun res (ty1,ty2) -> res >>= fun sol -> solve_eq sol ty1 ty2)
   | (TyFun (tys1, ty1), TyFun (tys2, ty2)) ->
      Error (Failure ("the function types's arugment are not different length"))
   | (TyUnion (ty11, ty12), TyUnion (ty21, ty22)) ->
   (* TODO: equality of unions *)
      Error (Failure ("not implemented: solve_eq with union types"))
   | (TyConstraint (ty1, c1), TyConstraint (ty2, c2)) ->
-     (* TODO: equality with constraint *)     
+     (* TODO: equality with constraint *)
      Error (Failure ("not implemented: solve_eq with constraint"))
   | _ ->
      Error (Failure (Printf.sprintf "must not eq type: %s =? %s" (show_typ ty1) (show_typ ty2)))
 
-let solve_sub sol ty1 ty2 = Ok sol (*TODO*)
+(* τ_1 ⊓ τ_2 *)
+let rec meet sol ty1 ty2 =
+  match (ty1, ty2) with
+  (* var *)
+  | TyVar x, _ ->
+     let ty1 = Option.value (Map.find sol x) ~default:TyAny in
+     meet sol ty1 ty2
+  | _, TyVar y ->
+     let ty2 = Option.value (Map.find sol y) ~default:TyAny in
+     meet sol ty1 ty2
+  (* any *)
+  | TyAny, _ -> ty2
+  | _, TyAny -> ty1
+  (* none *)
+  | TyNone, _ -> TyNone
+  | _, TyNone -> TyNone
+  (* struct *)
+  | TyStruct tys1, TyStruct tys2 when List.length tys1 = List.length tys2 ->
+     List.zip_exn tys1 tys2
+     |> List.map ~f:(fun (ty1, ty2) -> meet sol ty1 ty2)
+     |> (fun tys -> TyStruct tys)
+  (* function *)
+  | TyFun (args1, body1), TyFun (args2, body2) when List.length args1 = List.length args2 ->
+     List.zip_exn args1 args2
+     (* NOTE: using `meet` is the same as type derivation [ABS]. perhaps it should be `join` *)
+     |> List.map ~f:(fun (arg1, arg2) -> meet sol arg1 arg2)
```

Comment:
I think this should be `join` (`arg1 ⊔ arg2`), but Dialyzer employ `meet`. 
https://github.com/erlang/otp/blob/d6285b0a347b9489ce939511ee9a979acd868f71/lib/hipe/cerl/erl_types.erl#L2847

In order to make the same behavior of Dialyzer, I used `meet`.